### ISSUE TITLE
Thoughts on uploading contents from ImageBitmap to WebGPU

### ISSUE THREAD

**Issue by shaoboyan** [gpuweb/gpuweb#700]
`CopyImageBitmapToTexture` is the API used by WebGPU to interact with some HTMLElements like `<video>, <image>` in current spec. It uses `ImageBitmap` as the snapshot.
During the implementation of this API in Chromium, I find it is hard to achieve the best performance when interacting `ImageBitmap` with WebGPU Texture provided by users.  

Here are current status:

## Generate ImageBitmap
An `ImageBitmap` is an opaque container of resources. It can be backed by GPU memory or non-GPU memory. Users have limited info of this container (only width, height) and limited method to config it (through [ImageBitmapOptions](https://wiki.whatwg.org/wiki/ImageBitmap_Options)).

## Implicit steps in CopyImageBitmapToTexture
So when users invoke `copyImageBitmapToTexture`, they're in a great chance to hit:
- **color formats are not compatible** : ImageBitmap is a opaque container and users cannot know the accurate color formats. In WebGPU, we have a very strict rules in T2T copy. User can(and in a great chance to) provide a different format texture with ImageBitmap backed texture(if it is GPU backed). To resolve this, we need to start a rendering pipeline to do the conversion if dest format is renderable or read contents back to CPU and do conversion to cover more cases.
-   **transform** : ImageBitmap has options `imageOrientation = none`, `premultiplyAlpha = default` and `colorspaceConversion = default` in beginning.  The value `default` means *Implementation-specific behavior* . 
So we need to handle these transforms too. (For example, in Chromium, imageBitmap backed texture are Y flipped in implementation by default).
- **browser implementation related issue** : Except *Implementation-specific behavior* in ***transform*** part, there are other issues related with browser implementation. For example, SharedImage is the system used by chromium to share resources between contexts. In current chromium, ImageBitmap always backed by gl resource(or CPU buffer) and cannot be shared to dawn texture. We need an extra copy to handle it.

 
## Thoughts
Based on above facts, when users upload content from some elements to WebGPU texture through generating ImageBitmap and copyImageBitmapToTexture, they have great chance to trigger several transformations(or copies). First, the browser will do a snapshot first to generate a ImageBitmap. Second if the backed resource cannot upload to provided WebGPU texture directly, another pipeline(or CPU-side conversion) is needed to do the conversion. 

I have some thoughts on these steps and think maybe we can improve this through:
- **Interact with Video and Image element directly**: Current uploading model relies on snapshot step that generated `ImageBitmap`. In Chromium, these YUV format backed resources will do a YUV -> intermediate  renderable format(e.g. RGBA8Unorm)  conversion during snapshot and then follow CopyImageBitmapToTexture progress.  If we have an API that can interact with such elements backed by multi-planars directly (e.g. CreateTextureFromVideo), we can fully control this path and have more opportunity to accelerate it. For example, we can do YUV to dest_texture conversion directly and bypass all intermediate steps.

- **Add an option to ImageBitmap to notify it will be used by WebGPU** :  In Chromium, ImageBitmap that contains GPU resource is backed by non-sharable gl resource by default. And for perf consideration, this will still be the default path. But if ImageBitmap knows it will be used by WebGPU, it can be optimized. What's more, ImageBitmap defined lots of default options as "*Implementation-specific behavior*", so maybe this will help on more WebGPU specific optmization.

- **Texture format compatibility** : In current T2T copy rules in WebGPU, we don't have compatible format definition. So we need to start a pipeline to do some format conversion like RGBA8Unorm to RGBA8Uint. If we can use copy but not a pipeline, we may have perf improvement.
I did some rough exploring about format compatiblility. In Vulkan and D3D12, there are  rules to define compatibility. But metal is an open because it has strict rules for T2T copy and the possible compatible copy needs to leverage create texture view on a texture, which I have concern about the perf.

**Comment by kainino0x** [comment]
Unfortunately due to the nature of ImageBitmap, we'll have to automatically take care of a number of conversions. But it's important that it's possible to do no more work than necessary - so we need the API surface to allow us to avoid unnecessary multiple conversions or copies.

In the case of video frames, I think it's technically possible to keep the ImageBitmap in the source format, instead of converting. My understanding is this is in line with the new Web Codecs work as well, and fits with the concept of ImageBitmap. However, of course, there are more issues with sourcing data from videos - in particular that we want to avoid copying out of the decode queue unnecessarily - that will also need to be informed by progress on Web Codecs.

> * In Chromium, ImageBitmap that contains GPU resource is backed by non-sharable gl resource by default. And for perf consideration, this will still be the default path.

I realize this is an issue in our implementation, but in theory if we added a "forWebGPUDevice" to createImageBitmap, wouldn't it just push any conversion or api-transfer earlier? Would we not ultimately need to also put "forWebGPUDevice" on canvas elements, videos, etc, in order to avoid copying from gl-only textures to interop textures?

**Comment by AshleyScirra** [comment]
I'd add that a key purpose of ImageBitmap is to *asynchronously* prepare images for texture upload - `createImageBitmap` is async and in theory does all the conversion/premultiplication/etc up front (and ideally in parallel). If passing ImageBitmap to an API has to synchronously convert it, that's when jank happens. This was a problem with canvas2d's drawImage() in the past which would jank the first time an image was drawn; it would be a shame to bring back that problem in WebGPU.

**Comment by shaoboyan** [comment]
@AshleyScirra thanks for pointing the important part of ImageBitmap on **async**. I've focused more on **snapshot** part and forgot that.

And for @kainino0x's question,  to add options(e.g. "forWebGPUDevice"") in ImageBitmap, I think it can give hints to the source producer to generate correct back resource(through correct resource provider in Chromium).

According to the comments above, I think:
- An new attribute in ImageBitmap option (I think I'll call it `usage`) can help browser to opt this uploading path.
- For the video element case I cared more about, if the ImageBitmap contains "forWebGPUDevice" only, we can keep the raw multi-planars data from decoder in ImageBitmap. Since there must be a color conversion, we can combine all conversions in one during uploading progress.

And there is a more aggressive thought in my mind :
Inspired by GetPreferredTextureFormat in GPUSwapChain, if developers are not sensitive to the accurate texture format or other info.
We can have :
- `GetPreferredTextureFormat(ImageBitmap or HTMLElement)` to give a preferred format that helps to create a texture to avoid format conversion.
- Or `DescribeOfTextureMetadata ProduceGPUTextureFrom(ImageBitmap)` to generate a GPUTexture directly. In this path, it can has no copy but only resource sharing.

But there are concerns so I don't put them into the first comment because : 
- It may leak browser specific implementation details. So it may have security issue (Not sure)
- The assumption "developers are not sensitive to the accurate texture format or other info" means they knows these info in runtime and may need to prepare several shaders to handle this texture. (I don't know whether it is OK).

Raise them here for more thoughts on this.

**Comment by Kangz** [comment]
If it's doable, it would be very nice to have `ImageBitmaps` provide metadata on the best way to use their content, this way the application could work with the browser to reach a path that's as close to zero-copy as possible. `forWebGPUDevice` and a hint to the resource producer that further IBs might be used for WebGPU would work. That said IBs aren't specified in this forum and I'm sure other people would have opinions on this (like @fserb).

More random comments:

> It may leak browser specific implementation details. So it may have security issue (Not sure)

That's not a security concern, and should provide very small fingerprinting information, basically a subset of the bits that the data of your browser / OS combo would give.

> For the video element case I cared more about, if the ImageBitmap contains "forWebGPUDevice" only, we can keep the raw multi-planars data from decoder in ImageBitmap. Since there must be a color conversion, we can combine all conversions in one during uploading progress.

That's very related to https://github.com/gpuweb/gpuweb/issues/625, and there's two potential solutions that seem workable (and non-exclusive): having the video return the raw planes and metadata, or having an opaque "video texture" similarly to WebGL.

**Comment by RafaelCintron** [comment]
Chromium only: `forWebGPU` will help in the narrow case where D3D11 textures need to be shared with D3D12 textures on the same adapter, with no copies.  As systems transition off of D3D11 and onto D3D12, `forWebGPU` will not be necessary for same adapter usage.

However, if, in the future, we enable WebGPU on a different adapter as the rest of the browser, `forWebGPU`, by itself, doesn't provide enough information. You need to tell the browser which WebGPU adapter you are planning on using the texture so that the browser can upload the texture to the correct adapter to do async conversions. 

To avoid synchronous `ImageBitmap` conversions, I think we'll to add creation parameters for the texture format(s) and destination WebGPU adapter.

**Comment by AshleyScirra** [comment]
If there's going to be a WebGPU method to create an ImageBitmap, which you then pass to `copyImageBitmapToTexture`, would it make sense to just have a single async `uploadToTexture` method that accepts anything `createImageBitmap` would have?

**Comment by shaoboyan** [comment]
@AshleyScirra I don't think we need a method in WebGPU to create an ImageBitmap. I think what we might need is to let ImageBitmap be created with more metadata (Handled by browser to benefit WebGPU to use it) and a possible way to just resolve a texture(unlike copyImageBitmapToTexture, user provided a created texture) from a given ImageBitmap.

**Comment by Kangz** [comment]
+CC @beaufortfrancois @sandersdan

**Comment by greggman** [comment]
In WebGL2 I can upload a 2D image as slices of a 3D image. For example a 256x256 image can be uploaded as a 16x16x16 3D texture. Is that possible here?

I can also get sub rects from an image using the unpack parameters.

Would really love it if I could get sub rects in one go. Meaning, using ImageBitmap if I want to pull 16 different sub rects from an image I need to make 16 ImageBitmaps. I can't imagine that is performant. I'd rather be able to async load the image with some info on what format I want, then pull out sub rects out of that converted format.

Maybe there is a better interface but the WebGL1 spec says that for texImage2D that takes an ImageBitmap (or HTMLImageElement etc) as input

> The source image data is conceptually first converted to the data type and format specified by the format and type arguments

So I kind of just wish there as a sync or async API for that function as in

```
const arrayBuffer = await getImageBitmapAsData(srcImageBitmap, gl.RGBA, gl.UNSIGNED_BYTE);
```

Note: I get `gl.RGBA` and `gl.UNSIGNED_BYTE` are WebGL constants. Replace them with some other constants that mean the same thing.

Part of that desire is the code to do this is already there. It's just not exposed to JavaScript. Another is lots of devs including myself often want the data not for a texture but for something else. As it is there is no way short of including your own image decoders which sucks knowing all the code you need is only an missing API call away. Something that could be 1 line instead requires a multi 1000 line library.

I'm not saying you don't also need some opaque method to upload textures (like for video) but I am saying that I'd really like to see the current WebGL2 uses cases possible in WebGPU as well without having to decode images myself. And further, it seems like an opportunity to expose an API to just get the data.

**Comment by magcius** [comment]
@greggman this use case seems unrelated. You should be able to draw an ImageBitmap to a canvas using drawImage, and then getImageData on the canvas to decode the image data if necessary.

**Comment by greggman** [comment]
No, you can't. Canvas is premultplied alpha wihich means it's lossy if there is any alpha

**Comment by shaoboyan** [comment]
Hi, @greggman :
Try to provide my understanding of the cases in your comments and more info about this CopyImageBitmapToTexture API:

1. CopyImageBitmapToTexture accept three parameters: 
- ImageBitmapCopyView which contains an imageBitmap and the user set copy origin data,

- TextureCopyView which contains a webgpu texture and the user set  mipLevel, arrayLayer and origin.
- Extent: which contains user set width, height and depth.

2. About the case :
   > In WebGL2 I can upload a 2D image as slices of a 3D image. For example a 256x256 image can 
   be uploaded as a 16x16x16 3D texture. Is that possible here?

   My understanding is that for this case in WebGL, the workflow should be:
   `ReadPixels ----->PackedBuffer---->TexImage3D with unpacked buffer`
   And it will cost two GL commands to do the uploading.
   While I think in WebGPU it could be:
   `CopyImageBitmapToTexture ----> T2B copy ---> B2T copy to upload to 3D Image` 
   It will cost 3 commands to do the uploading.

3. I think the one of points is how to deal with the sub-rect uploading or downloading from 
    ImageBitmap with an user defined texture format.
    
    For sub-rect uploading, I think CopyImageBitmapToTexture has the ability to upload one sub-rect 
    content of ImageBitmap(not accurate) to a layer of a webgpu texture.

    For downloading, as the case:
    > Meaning, using ImageBitmap if I want to pull 16 different sub rects from an image I need to make 16 ImageBitmaps.
     
   I think it should be:
   `CopyImageBitmapToTexture -----> T2B copy with whole imageBimap or several T2B copies with subrects you want ----->map buffers to read contents`

4. My understand of the API `getImageBitmapAsData(srcImageBitmap, gl.RGBA, gl.UNSIGNED_BYTE)` is that developers want to let browser handle the decode(by creating ImageBitmap) and conversion(color format, orientation and so on) and return the final result.


It is similar to CopyImageBitmapToTexture(user provided the texture) but it cannot reading contents(to CPU) directly.

**Comment by kainino0x** [comment]
@greggman, does Shaobo's suggestion (one copyIB2T + multiple copyT2B or copyT2T calls) resolve your issue? Or do you expect cases where that initial copyIB2T would copy much more data than you actually need?

**Comment by Kangz** [comment]
Closing this issue since we have `copyExternalImageToTexture`. Further discussion should be in new issues.
